### PR TITLE
[Merged by Bors] - feat(Nat/GCD): Add dvd_gcd_mul_gcd_of_dvd_mul

### PR DESCRIPTION
**Pull request by Jun2M**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Nat/GCD/Basic.lean [hunk 1]
```diff
@@ -345,4 +345,26 @@ theorem Coprime.mul_add_mul_ne_mul {m n a b : ℕ} (cop : Coprime m n) (ha : a 
...
 
+theorem dvd_gcd_mul_of_dvd_mul {x n m : ℕ} (h : x ∣ n * m) : x ∣ gcd x n * m := by
+  rcases h with ⟨ y, hy ⟩
+  rw [← gcd_mul_right, hy, gcd_mul_left]
+  exact dvd_mul_right x (gcd m y)
+
+theorem dvd_mul_gcd_of_dvd_mul {x n m : ℕ} (h : x ∣ n * m) : x ∣ n * gcd x m := by
+  rcases h with ⟨ y, hy ⟩
+  rw [← gcd_mul_left, hy, mul_comm, gcd_mul_left]
+  exact dvd_mul_right x (gcd n y)
```
Comment (erdOne) [inline Mathlib/Data/Nat/GCD/Basic.lean:356]: I think it would be useful to make them iffs.

File: Mathlib/Data/Nat/GCD/Basic.lean [hunk 2]
```diff
@@ -345,4 +345,26 @@ theorem Coprime.mul_add_mul_ne_mul {m n a b : ℕ} (cop : Coprime m n) (ha : a 
   rw [← mul_assoc, ← h, add_mul, add_mul, mul_comm _ n, ← mul_assoc, mul_comm y]
 #align nat.coprime.mul_add_mul_ne_mul Nat.Coprime.mul_add_mul_ne_mul
 
+theorem dvd_gcd_mul_of_dvd_mul {x n m : ℕ} (h : x ∣ n * m) : x ∣ gcd x n * m := by
+  rcases h with ⟨ y, hy ⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Nat/GCD/Basic.lean:349]: Style:
```suggestion
  rcases h with ⟨y, hy⟩
```
(also below)

File: Mathlib/Data/Nat/GCD/Basic.lean (hunk 1 above)
Comment (Jun2M) [reply Mathlib/Data/Nat/GCD/Basic.lean]: Done!

File: Mathlib/Data/Nat/GCD/Basic.lean [hunk 3]
```diff
@@ -345,4 +345,49 @@ theorem Coprime.mul_add_mul_ne_mul {m n a b : ℕ} (cop : Coprime m n) (ha : a 
...
+
+theorem dvd_gcd_mul_of_dvd_mul  (h : x ∣ n * m) : x ∣ gcd x n * m := by
+  obtain ⟨y, hy⟩ := h
+  rw [← gcd_mul_right, hy, gcd_mul_left]
+  exact dvd_mul_right x (gcd m y)
+
+theorem dvd_mul_of_dvd_gcd_mul (h : x ∣ gcd x n * m) : x ∣ n * m := by
+  obtain ⟨y, hy⟩ := h
+  use y * (n / gcd x n)
+  rw [← mul_assoc, ← hy, mul_comm (_ * m), ← mul_assoc, Nat.div_mul_cancel (gcd_dvd_right x n)]
```
Comment (fpvandoorn) [inline Mathlib/Data/Nat/GCD/Basic.lean:358]: Prove this using `apply h.trans` and `mul_dvd_mul_right` and [Nat.gcd_dvd_right](https://leanprover-community.github.io/mathlib4_docs/Init/Data/Nat/Gcd.html#Nat.gcd_dvd_right).

File: Mathlib/Data/Nat/GCD/Basic.lean [hunk 4]
```diff
@@ -345,4 +345,49 @@ theorem Coprime.mul_add_mul_ne_mul {m n a b : ℕ} (cop : Coprime m n) (ha : a 
...
+
+theorem dvd_gcd_mul_iff_dvd_gcd : x ∣ gcd x n * m ↔ x ∣ n * m :=
```
Comment (fpvandoorn) [inline Mathlib/Data/Nat/GCD/Basic.lean:360]: I think you should remove the previous two lemmas and inline their proofs here.

File: Mathlib/Data/Nat/GCD/Basic.lean [hunk 5]
```diff
@@ -345,4 +345,49 @@ theorem Coprime.mul_add_mul_ne_mul {m n a b : ℕ} (cop : Coprime m n) (ha : a 
...
+  obtain ⟨y, hy⟩ := h
+  rw [← gcd_mul_left, hy, mul_comm, gcd_mul_left]
+  exact dvd_mul_right x (gcd n y)
+
+theorem dvd_mul_of_dvd_mul_gcd (h : x ∣ n * gcd x m) : x ∣ n * m := by
+  obtain ⟨y, hy⟩ := h
+  use y * (m / gcd x m)
+  rw [← mul_assoc, ← hy, mul_assoc, mul_comm (gcd x m), Nat.div_mul_cancel (gcd_dvd_right x m)]
+
+theorem dvd_mul_gcd_iff_dvd_gcd : x ∣ n * gcd x m ↔ x ∣ n * m :=
```
Comment (fpvandoorn) [inline Mathlib/Data/Nat/GCD/Basic.lean:373]: remove the previous two lemmas, and prove this using `dvd_gcd_mul_iff_dvd_gcd` and `mul_comm`.

File: Mathlib/Data/Nat/GCD/Basic.lean (hunk 4 above)
Comment (fpvandoorn) [reply Mathlib/Data/Nat/GCD/Basic.lean:360]: Rename to `dvd_gcd_mul_iff_dvd_mul` (also below)

File: Mathlib/Data/Nat/GCD/Basic.lean [hunk 6]
```diff
@@ -345,4 +345,49 @@ theorem Coprime.mul_add_mul_ne_mul {m n a b : ℕ} (cop : Coprime m n) (ha : a 
...
+    gcd x n * gcd x m = x := by
+  apply dvd_antisymm
+  · apply (hcop.gcd_both x x).mul_dvd_of_dvd_of_dvd <;> exact gcd_dvd_left x _
+  · exact dvd_gcd_mul_gcd_iff_dvd_mul.mpr h
+
+theorem dvd_mul_of_gcd_mul_gcd_eq (h : gcd x n * gcd x m = x) : x ∣ n * m := by
+  rw [← h]
+  apply Nat.mul_dvd_mul <;> exact gcd_dvd_right _ _
+
+theorem gcd_mul_gcd_eq_iff_dvd_mul_of_coprime (hcop : Coprime n m) :
```
Comment (fpvandoorn) [inline Mathlib/Data/Nat/GCD/Basic.lean:389]: again, inline the proofs of the previous two theorems here, please.

**Review by fpvandoorn — COMMENTED** [review]
Thanks for your PR! Generally if we prove an `iff` we don't want to state the individual implications separately (unless they hold under weaker assumptions - `dvd_mul_of_gcd_mul_gcd_eq` has weaker assumptions, but is probably not very useful by itself).

**Comment by fpvandoorn** [comment]
Thanks!

bors merge

**Pull request closed without merging**